service/worker/indexer/esProcessor.go

```diff
@@ -0,0 +1,265 @@
+			continue
+		}
+		responseItem := responseItems[i]
+		for _, resp := range responseItem {
+			key := resp.Id + visType
+			switch {
+			case isResponseSuccess(resp.Status):
+				p.ackKafkaMsg(key)
+			case !isResponseRetriable(resp.Status):
+				p.nackKafkaMsg(key)
+			default:
+				// do nothing, bulk processor will retry
+			}
+		}
+	}
+}
+
+func (p *esProcessorImpl) ackKafkaMsg(key string) {
+	p.ackKafkaMsgHelper(key, false)
+}
+
+func (p *esProcessorImpl) nackKafkaMsg(key string) {
+	p.ackKafkaMsgHelper(key, true)
+}
+
+func (p *esProcessorImpl) ackKafkaMsgHelper(key string, nack bool) {
+	msg, ok := p.mapToKafkaMsg.Get(key)
+	if !ok {
+		return // duplicate kafka message
+	}
+	kafkaMsg, ok := msg.(messaging.Message)
+	if !ok { // must be bug in code and bad deployment
+		p.logger.WithFields(bark.Fields{
+			logging.TagESKey: key,
+		}).Fatal("Message is not kafka message.")
+	}
+
+	if nack {
+		kafkaMsg.Nack()
+	} else {
+		kafkaMsg.Ack()
+	}
+	p.mapToKafkaMsg.Remove(key)
+}
+
+func (p *esProcessorImpl) hashFn(key interface{}) uint32 {
+	id, ok := key.(string)
+	if !ok {
+		return 0
+	}
+	numOfShards := p.config.IndexerConcurrency()
+	return uint32(common.WorkflowIDToHistoryShard(id, numOfShards))
+}
+
+func (p *esProcessorImpl) getVisibilityType(request elastic.BulkableRequest) string {
+	req, err := request.Source()
+	if err != nil {
+		p.logger.WithFields(bark.Fields{
+			logging.TagErr:       err,
+			logging.TagESRequest: request.String(),
+		}).Error("Get request source err.")
+		p.metricsClient.IncCounter(metrics.ESProcessorScope, metrics.ESProcessorCorruptedData)
+		return ""
+	}
+
+	var reqHead map[string]map[string]interface{}
+	if err := json.Unmarshal([]byte(req[0]), &reqHead); err != nil {
+		p.logger.WithFields(bark.Fields{
+			logging.TagErr: err,
+		}).Error("Unmarshal request err.")
+		p.metricsClient.IncCounter(metrics.ESProcessorScope, metrics.ESProcessorCorruptedData)
+		return ""
+	}
+
+	var visType string
+	for _, header := range reqHead {
+		h, ok := header["version"].(float64)
+		if !ok {
+			p.logger.Error("Request missing version field.")
+			p.metricsClient.IncCounter(metrics.ESProcessorScope, metrics.ESProcessorCorruptedData)
+			return ""
+		}
+		visType = p.convertESVersionToVisibilityType(h)
+	}
+	return visType
+}
+
+func (p *esProcessorImpl) convertESVersionToVisibilityType(version float64) string {
+	switch version {
+	case float64(versionForOpen):
+		return indexer.VisibilityMsgTypeOpen.String()
+	case float64(versionForClose):
+		return indexer.VisibilityMsgTypeClosed.String()
+	case float64(versionForDelete):
+		return indexer.VisibilityMsgTypeDelete.String()
+	}
+
+	p.logger.Errorf("Unknown version %v in ES request. ", version)
+	p.metricsClient.IncCounter(metrics.ESProcessorScope, metrics.ESProcessorCorruptedData)
+	return ""
+}
+
+// 409 - Version Conflict
+// 404 - Not Found
+func isResponseSuccess(status int) bool {
+	if status >= 200 && status < 300 || status == 409 || status == 404 {
```

Comment:
Why '404' is treated as success? Won't this result is loosing messages?

Reply:
404 can be returned by delete requests (not yet added). It's treated as success, since the doc to delete not found.